src/gui/SampleThumbnail.cpp

```diff
@@ -0,0 +1,216 @@
+{
+	const auto lengthY1 = bit.max * scalingFactor;
+	const auto lengthY2 = bit.min * scalingFactor;
+
+	const auto lineY1 = centerY - lengthY1;
+	const auto lineY2 = centerY - lengthY2;
+
+	const auto innerLineY1 = centerY - bit.max*0.6 * scalingFactor;
+	const auto innerLineY2 = centerY - bit.min*0.6 * scalingFactor;
+
+	painter.drawLine(QPointF{lineX, lineY1}, QPointF{lineX, lineY2});
+	painter.setPen(innerColor);
+
+	painter.drawLine(QPointF{lineX, innerLineY1}, QPointF{lineX, innerLineY2});
+	painter.setPen(color);
+}
+
+void SampleThumbnail::visualize(const VisualizeParameters& parameters, QPainter& painter) const
+{
+	const auto& clipRect = parameters.clipRect;
+	const auto& sampRect = parameters.sampRect.isNull() ? clipRect : parameters.sampRect;
+	const auto& drawRect = parameters.drawRect.isNull() ? clipRect : parameters.drawRect;
+
+	const auto sampleViewLength = parameters.sampleEnd - parameters.sampleStart;
+
+	const auto x = sampRect.x();
+	const auto height = clipRect.height();
+	const auto halfHeight = height / 2;
+	const auto width = sampRect.width();
+	const auto centerY = clipRect.y() + halfHeight;
+
+	if (width < 1) { return; }
+
+	const auto scalingFactor = halfHeight * parameters.amplification;
+
+	const auto color = painter.pen().color();
+	const auto innerColor = color.lighter(123);
+
+	const auto widthSelect = static_cast<std::size_t>(static_cast<float>(width) / sampleViewLength);
+
+	auto thumbnailIt = m_thumbnailCache->end();
+
+	do
+	{
+		thumbnailIt--;
+	}
+	while (thumbnailIt != m_thumbnailCache->begin() && thumbnailIt->size() < widthSelect);
+
+	const auto& thumbnail = *thumbnailIt;
+
+	const auto thumbnailLastSample 	= std::max<size_t>(static_cast<size_t>(parameters.sampleEnd * thumbnail.size()), 1) - 1;
+	const auto tViewStart 			= static_cast<long>(parameters.sampleStart * thumbnail.size());
+	const auto tViewLast 			= std::min<size_t>(thumbnailLastSample, thumbnail.size() - 1);
+	const auto tLast				= thumbnail.size()-1;
+	const auto thumbnailViewSize 	= thumbnailLastSample + 1 - tViewStart;
+
+	const auto pixelIndexStart 	= std::max(std::max(clipRect.x(), drawRect.x()), x);
+	const auto pixelIndexEnd 	= std::min(std::min(clipRect.width(), drawRect.width()), width) + pixelIndexStart;
+
+	const auto tChunk = (thumbnailViewSize + width) / width;
+
+	for (auto pixelIndex = pixelIndexStart; pixelIndex <= pixelIndexEnd; pixelIndex++)
+	{
+		auto tIndex = tViewStart + (pixelIndex - x) * thumbnailViewSize / width;
+
+		if (tIndex > tViewLast) { break; }
+
+		const auto tChunkBound = tIndex + tChunk;
+
+		auto thumbnailBit = Bit{};
+
+		if (parameters.reversed)
+		{
+			thumbnailBit = std::accumulate(
+					thumbnail.cbegin() + tLast - tChunkBound,
+					thumbnail.cbegin() + tLast - tIndex,
+					Bit{}
+				);
+		}
+		else
+		{
+			thumbnailBit = std::accumulate(
+					thumbnail.cbegin() + tIndex,
+					thumbnail.cbegin() + tChunkBound,
+					Bit{}
+				);
+		}
```

Comment:
You can probably just calculate the correct indices (i.e., reversed or not) within the loop and then just call `std::accumulate`.